---
layout: post
title: 8_nnet问题解决记录
category: app
typora-root-url: ../../..
---



1. cgo调用c层回调函数，崩溃堆栈无法定位问题。通过尝试各种选项，效果如下：

   * 不使用调试信息裁剪这个编译项，即去除`-ldflags="-s -w"`（其中-w会去掉gdb调试信息）。则使用gdb调试即可打印golang堆栈，但是c层堆栈只能显示出最后一层。这样也能看出哪行代码崩溃的！
   * 使用`-ldflags="-s -w"`，则gdb无法看到golang堆栈，但也能看出哪行代码崩溃的。
   * 使用dlv调试，可以看出C层崩溃堆栈，但是dlv无法加载golang动态库，所以无法正常显示golang堆栈。
   * golang崩溃堆栈，在任何编译项下，都能在崩溃时正常输出到标准错误输出。因为默认`GOTRACEBACK=single`

   综上，最佳实践是：

   * 使用`-ldflags="-s -w"`使库文件大小尽可能小。
   * 出现崩溃，需要记录标准错误输出，这里可以使用重定向标准错误输出的方式将崩溃堆栈重定向到文件，程序发生崩溃后，直接查看对应文件即可获得崩溃堆栈。

2. 期望通过主动轮询收包队列，将数据包接收到同个线程，避免加锁。
   背景：当前是收到包后，主动通过回调函数将数据包通知给使用方，最大的问题是”回调函数执行者是一个新的线程“，使用方往往期望能在调用接口的线程接收到函数，否则使用方需要自己加锁。
   实现：使用无锁ringbuffer（https://github.com/hedzr/go-ringbuf.git），将接收到的数据包存入该缓存；提供`xxxRecv()`接口，同步地从该缓存读取数据包。

3. 客户端rcvwnd设置过小，暴露出服务端重启后，阻塞在发送数据处无法接收新请求！

   原因：https://github.com/hedzr/go-ringbuf.git

4. 压测效率较低。



## 上线应用篇

- **Q：**异常协议包导致崩溃。
  **A：**可以通过构造异常包发包进行测试。构造异常包可以使用scapy来快速构造，编写如下脚本：

  ```python
  #coding: utf-8
  #filename: udp_pkt_scapy.py
  
  import sys
  import struct
  import platform as sysPlatform
  from scapy.all import *
  
  # 支持loopback发包. ref: https://scapy.readthedocs.io/en/latest/troubleshooting.html#my-tcp-connections-are-reset-by-scapy-or-by-my-kernel
  if sysPlatform.system().lower() == "linux":
  	conf.L3socket = L3RawSocket
  
  # 构造协议包
  payloadLst = [0, 0, 0, 0, 241, 0, 52, 0, 125, 6, 20, 34, 85, 0, 0, 4, 6, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 6, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 104, 101, 108, 108, 111, 0]
  payload = struct.pack("B"*len(payloadLst), *payloadLst)
pkt = IP(dst=sys.argv[1])/UDP(dport=int(sys.argv[2]))/bytes(payload)
  send(pkt)
  ```
```
  
  运行脚本进行测试：
  
​```shell
  sudo python udp_pkt_scapy.py 1.1.1.1 12345
```

  另外，如果是python2的话，应该安装scapy==2.4.5版本，否则执行时会报错。

- **Q：**使用了golang库后，SetUnhandledExceptionFilter被屏蔽。
  **A：**这里要区分x64和x86平台：
  	1. 对于x64平台，go1.22之前并不支持触发SetUnhandledExceptionFilter的调用，后续[#50951](https://github.com/golang/go/issues/50951)提出并修复了该问题。
   	2. 对于x86平台，一直以来是支持触发SetUnhandledExceptionFilter的调用的，不过golang内部对于`windows-386`会调用SetUnhandledExceptionFilter用于信号处理，屏蔽了cpp层设置的处理函数，可以使用[detours](https://www.microsoft.com/en-us/research/project/detours/)工具排查。解决方案也有两个：
       （1）使用网上建议的[DisableSetUnhandledExceptionFilter](https://blog.csdn.net/weixin_41504987/article/details/131461767)代码禁止后续SetUnhandledExceptionFilter覆盖，本质上是通过修改自己进程的内存来实现。
       （2）修改golang源码，不在信号处理中调用SetUnhandledExceptionFilter



- **Q：**nnet踢用户流程梳理。
  **A：**梳理如下：
  
- 

- **Q：**win平台下udp连接调用`udpconn.RecvFrom()`返回错误：
  **A：**梳理如下：



- **Q：**ios运行崩溃，前面日志有出现“Too many open files”错误。
  **A：**原因：ios进程（arm64架构）可打开的文件描述符数量为255，如果打开文件描述符超过了，相关的创建socket、打开文件等操作就会失败，如果没有处理好就容易导致崩溃。
  复现方法：
  1、使用`dup(STDIN_FILENO)`占用255个文件描述符。
  2、尝试执行创建socket操作触发崩溃。

  检查文件描述符数量：

  1、打开Instruments的File Activity分析：Xcode13.4 - Services - File Activity

  2、上方选中具体需要测试的设备 - 相应进程，然后点击开始录制。

  3、此时进程会自动重启，然后开始执行测试操作。

  4、操作完毕，再回到File Activity中，点击左上方停止录制。

  5、在File Activity中，下侧面板左上方选中“File Descriptor History”，就可以显示进程所使用的文件描述符数量了。

  修复方式：

  1、创建socket如果返回错误err，就不要继续使用空结构体。

  